lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemStats.java

```diff
@@ -0,0 +1,222 @@
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getPutObject()
+    {
+        return putObject;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getCreateMultipartUpload()
+    {
+        return createMultipartUpload;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getCompleteMultipartUpload()
+    {
+        return completeMultipartUpload;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getAbortMultipartUpload()
+    {
+        return abortMultipartUpload;
+    }
+
+    public MetricPublisher newMetricPublisher()
+    {
+        return new JmxMetricPublisher(this);
+    }
+
+    public final class JmxMetricPublisher
+            implements MetricPublisher
+    {
+        private static final Logger log = Logger.get(JmxMetricPublisher.class);
+
+        private final S3FileSystemStats stats;
+
+        public JmxMetricPublisher(S3FileSystemStats stats)
+        {
+            this.stats = requireNonNull(stats, "stats is null");
+        }
+
+        private static final Set<SdkMetric<?>> ALLOWED_METRICS = Set.of(API_CALL_SUCCESSFUL, RETRY_COUNT, API_CALL_DURATION, ERROR_TYPE);
+
+        private static final Set<String> ALLOWED_SERVICE_IDS = Set.of("S3");
+
+        public void publish(MetricCollection metricCollection)
+        {
+            try {
+                Optional<String> serviceId = metricCollection.metricValues(SERVICE_ID).stream().filter(Objects::nonNull).findFirst();
+                Optional<String> operationName = metricCollection.metricValues(OPERATION_NAME).stream().filter(Objects::nonNull).findFirst();
+                if (serviceId.isEmpty() || operationName.isEmpty()) {
+                    log.warn("ServiceId or OperationName is empty for AWS MetricCollection: %s", metricCollection);
+                    return;
+                }
+
+                if (!ALLOWED_SERVICE_IDS.contains(serviceId.get())) {
+                    return;
+                }
+
+                AwsSdkV2ApiCallStats apiCallStats = getApiCallStats(operationName.get());
+                publishMetrics(metricCollection, apiCallStats);
+            }
+            catch (Exception e) {
+                log.warn(e, "Publishing AWS metrics failed");
+            }
+        }
+
+        private void publishMetrics(MetricCollection metricCollection, AwsSdkV2ApiCallStats apiCallStats)
+        {
+            metricCollection.stream()
+                    .filter(metricRecord -> metricRecord.metric() != null && metricRecord.value() != null)
+                    .filter(metricRecord -> ALLOWED_METRICS.contains(metricRecord.metric()))
+                    .forEach(metricRecord -> {
+                        if (metricRecord.metric().equals(API_CALL_SUCCESSFUL)) {
+                            Boolean value = (Boolean) metricRecord.value();
+
+                            increment(stats.total.getCalls());
+                            if (apiCallStats != null) {
```

Comment:
Let's make `getApiCallStats()` return a dummy instance for unknown operations so that we don't need all the null checks. That will make the code easier to read.

Reply:
uploadPart